documentation/specs/src/base-ledger/replay-protection.md

```diff
@@ -0,0 +1,100 @@
+
+## WrapperTx
+
+`WrapperTx` is the only type of transaction currently accepted by the ledger. It must be protected from replay attacks because, if it wasn't, a malicious user could replay the transaction as is. Even if the inner transaction implemented replay protection or, for any reason, wasn't accepted, the signer of the wrapper would still pay for gas and fees, effectively suffering economic damage.
+
+To protect this transaction we can implement an in-protocol mechanism that requires us to expand the current definition of the struct to include a transaction counter:
+
+```rust
+pub struct WrapperTx {
+    /// The fee to be payed for including the tx
+    pub fee: Fee,
+    /// Used to determine an implicit account of the fee payer
+    pub pk: common::PublicKey,
+    /// The epoch in which the tx is to be submitted. This determines
+    /// which decryption key will be used
+    pub epoch: Epoch,
+    /// Max amount of gas that can be used when executing the inner tx
+    pub gas_limit: GasLimit,
+    /// Transaction counter for replay-protection
+    pub tx_counter: u32,
+    /// the encrypted payload
+    pub inner_tx: EncryptedTx,
+    /// sha-2 hash of the inner transaction acting as a commitment
+    /// the contents of the encrypted payload
+    pub tx_hash: Hash,
+}
+``` 
+
+In addition, we need a counter in the storage subspace of every implict address:
+
+```
+/$Address/tx_counter: u32
+```
+
+In `process_proposal` we check that `tx_counter` field in `WrapperTx` is greater or equal to the value currently in storage; if this doesn't hold, the transaction is considered invalid.
+
+At last, in `finalize_block`, the protocol updates the counter key in storage, increasing its value by one. Now, if a malicious user tried to replay this transaction, the `tx_counter` in the struct would no longer be greater or equal to the one in storage and the transaction would be deemed invalid. 
+
+Note that the address whose counter is involved in this process is the one signing the transaction (the same as the `pk` field of the struct).
+
+## InnerTx
+
+The `EncryptedTx` incapsulated inside `WrapperTx` should be protected too. That's because, if it wasn't, an attacker could extract the inner tx, rewrap it, and replay it.\
+To simplify this check we will perform it entirely in Wasm: the check of the counter will be carryed out by the validity predicates while the actual writing of the counter in storage will be done by the transactions themselves. 
+
+To do so, the `SignedTxData` attached to the transaction will hold the current value of the counter in storage. The transaction will simply read the value from storage and increase its value by one. The VP will then check the validity of the signature and, if it's deemed valid, will proceed to checking if the pre value of the counter in storage was less or equal than the one contained in the transaction data and if the post value of the key in storage has been incremented by one.
```

Comment:
same as above, I think this should be "checking if the pre value of the counter in storage was equal to the one contained in the transaction data"